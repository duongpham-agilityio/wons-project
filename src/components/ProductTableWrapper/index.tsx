import { getData } from '@/services';
import { ProductTable } from '@/components';

interface IProductTableWrapper {
  searchQuery: string;
  currentPage: number;
  filter?: string;
}

const itemsPerPage = 10;

const ProductTableWrapper = async ({
  searchQuery,
  currentPage,
  filter,
}: IProductTableWrapper) => {
  const allData = await getData(searchQuery);
  const safeData = allData || [];

  let filteredData = safeData;
  if (filter === 'views_below_5k') {
    filteredData = safeData.filter((item) => Number(item.views) < 5000);
  } else if (filter === 'views_5k_plus') {
    filteredData = safeData.filter((item) => Number(item.views) >= 5000);
  }

  // Pagination logic
  const indexOfLastItem = currentPage * itemsPerPage;
  const indexOfFirstItem = indexOfLastItem - itemsPerPage;
  const currentItems = filteredData.slice(indexOfFirstItem, indexOfLastItem);

  return (
    <ProductTable
      products={currentItems}
      currentPage={currentPage}
      itemsPerPage={itemsPerPage}
      totalItems={filteredData.length}
    />
  );
};

export default ProductTableWrapper;
