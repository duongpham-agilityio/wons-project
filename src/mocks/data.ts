const dateDicker = [
  { key: 'day', label: 'Day' },
  { key: 'weekly', label: 'Weekly' },
  { key: 'month', label: 'Month' },
  { key: 'year', label: 'Year' },
];

const productSelector = [
  { key: 'allProducts', label: 'All Products' },
  { key: 'products', label: 'Products' },
];

const data = [
  {
    value: '$39.510,32',
    label: 'Total Revenue',
    growth: '+12,7 %',
  },
  {
    value: '175.182',
    label: 'Total Sales',
    growth: '+5,1 %',
  },
  {
    value: '508.072',
    label: 'Customers',
    growth: '+18,3 %',
  },
  {
    value: '105.075',
    label: 'New Subscriptions',
    growth: '+10,6 %',
  },
];

const rows = [
  {
    key: '1',
    name: 'Basic design guideline',
    status: '9 Jan 2023 9:43 PM',
    views: '3.147',
    sales: '1.004',
    conversion: '6,5%',
    total: '$14.238',
  },
  {
    key: '2',
    name: 'Basic design guideline',
    status: '9 Jan 2023 9:43 PM',
    views: '3.147',
    sales: '1.004',
    conversion: '6,5%',
    total: '$14.238',
  },
  {
    key: '3',
    name: 'Basic design guideline',
    status: '9 Jan 2023 9:43 PM',
    views: '3.147',
    sales: '1.004',
    conversion: '6,5%',
    total: '$14.238',
  },
  {
    key: '4',
    name: 'Basic design guideline',
    status: '9 Jan 2023 9:43 PM',
    views: '3.147',
    sales: '1.004',
    conversion: '6,5%',
    total: '$14.238',
  },
];

const columns = [
  {
    key: 'no',
    label: 'No',
    className: 'w-20',
  },
  {
    key: 'img',
    label: '',
    className: 'w-16',
  },
  {
    key: 'name',
    label: 'Name',
  },
  {
    key: 'status',
    label: 'Status',
  },
  {
    key: 'views',
    label: 'Views',
  },
  {
    key: 'sales',
    label: 'Sales',
  },
  {
    key: 'conversion',
    label: 'Conversion',
  },
  {
    key: 'total',
    label: 'Total',
  },
  {
    key: 'details',
    label: 'Details',
  },
];

export { dateDicker, productSelector, data, rows, columns };
