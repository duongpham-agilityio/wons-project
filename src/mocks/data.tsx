import Link from 'next/link';
import { ROUTERS } from '@/constants';
import { ProductAction } from '@/types';

const dateDicker = [
  { key: 'day', label: 'Day' },
  { key: 'weekly', label: 'Weekly' },
  { key: 'month', label: 'Month' },
  { key: 'year', label: 'Year' },
];

const productSelector = [
  { key: 'allProducts', label: 'All Products' },
  { key: 'products', label: 'Products' },
];

const data = [
  {
    value: '$39,510.32',
    label: 'Total Revenue',
    growth: '+12,7 %',
  },
  {
    value: '175,182',
    label: 'Total Sales',
    growth: '+5,1 %',
  },
  {
    value: '508,072',
    label: 'Customers',
    growth: '+18,3 %',
  },
  {
    value: '105,075',
    label: 'New Subscriptions',
    growth: '+10,6 %',
  },
];

const rows = [
  {
    key: '1',
    name: 'Basic design guideline',
    status: '9 Jan 2023 9:43 PM',
    views: '3.147',
    sales: '1.004',
    conversion: '6,5%',
    total: '$14.238',
  },
  {
    key: '2',
    name: 'Basic design guideline',
    status: '9 Jan 2023 9:43 PM',
    views: '3.147',
    sales: '1.004',
    conversion: '6,5%',
    total: '$14.238',
  },
  {
    key: '3',
    name: 'Basic design guideline',
    status: '9 Jan 2023 9:43 PM',
    views: '3.147',
    sales: '1.004',
    conversion: '6,5%',
    total: '$14.238',
  },
  {
    key: '4',
    name: 'Basic design guideline',
    status: '9 Jan 2023 9:43 PM',
    views: '3.147',
    sales: '1.004',
    conversion: '6,5%',
    total: '$14.238',
  },
];

const columns = [
  {
    key: 'no',
    label: 'No',
    className: 'w-20',
  },
  {
    key: 'img',
    label: '',
    className: 'w-16',
  },
  {
    key: 'name',
    label: 'Name',
  },
  {
    key: 'status',
    label: 'Status',
  },
  {
    key: 'views',
    label: 'Views',
  },
  {
    key: 'sales',
    label: 'Sales',
  },
  {
    key: 'conversion',
    label: 'Conversion',
  },
  {
    key: 'total',
    label: 'Total',
  },
  {
    key: 'details',
    label: 'Details',
    className: 'w-10',
  },
  {
    key: 'dropdown',
    label: '',
  },
];

const items = [
  {
    key: 'views_below_5k',
    label: 'Views < 5000',
  },
  {
    key: 'views_5k_plus',
    label: 'Views â‰¥ 5000',
  },
];

const productActions: ProductAction[] = [
  {
    key: 'new',
    label: <Link href={ROUTERS.ADD_PRODUCT}>Add New Product</Link>,
    color: 'default',
  },
  {
    key: 'edit',
    label: <Link href={ROUTERS.EDIT_PRODUCT}>Edit Product</Link>,
    color: 'default',
  },
  {
    key: 'delete',
    label: 'Delete Product',
    className: 'text-danger',
    color: 'danger',
  },
];

export {
  dateDicker,
  productSelector,
  data,
  rows,
  columns,
  items,
  productActions,
};
