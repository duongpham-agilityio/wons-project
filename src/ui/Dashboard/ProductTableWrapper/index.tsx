import { filterProducts, paginateProducts } from '@/utils';
import { IProducts } from '@/types';
import { ProductTable } from '@/ui';

interface IProductTableWrapper {
  currentPage: number;
  filter?: string;
  data: IProducts[];
}

const itemsPerPage = 10;

const ProductTableWrapper = ({
  currentPage,
  filter,
  data,
}: IProductTableWrapper) => {
  const safeData = Array.isArray(data) ? data : [];

  const filteredData = filterProducts(safeData, filter);
  const currentItems = paginateProducts(
    filteredData,
    currentPage,
    itemsPerPage,
  );

  return (
    <ProductTable
      products={currentItems}
      currentPage={currentPage}
      itemsPerPage={itemsPerPage}
      totalItems={filteredData.length}
    />
  );
};

export default ProductTableWrapper;
