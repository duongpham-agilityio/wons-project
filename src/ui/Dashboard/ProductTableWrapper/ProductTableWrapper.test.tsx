import { render } from '@testing-library/react';
import { useRouter } from 'next/navigation';
import { rows } from '@/mocks';
import ProductTableWrapper from '.';

jest.mock('next/navigation', () => ({
  useRouter: jest.fn(),
}));

describe('ProductTableWrapper Component', () => {
  beforeEach(() => {
    (useRouter as jest.Mock).mockReturnValue({
      push: jest.fn(),
    });
  });

  test('render product table wrapper', () => {
    const { container } = render(
      <ProductTableWrapper currentPage={10} data={rows} />,
    );
    expect(container).toMatchSnapshot();
  });
});
