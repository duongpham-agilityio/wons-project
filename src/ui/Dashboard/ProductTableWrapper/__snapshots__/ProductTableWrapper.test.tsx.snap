// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductTableWrapper Component render product table wrapper 1`] = `
<div>
  <div
    class="flex flex-col relative gap-4 w-full mt-5"
  >
    <div
      class="p-4 z-0 flex flex-col relative justify-between gap-4 bg-content1 overflow-auto rounded-large shadow-small w-full py-7 px-4xl"
    >
      <table
        aria-describedby=""
        aria-label="Controlled table example with dynamic content"
        aria-multiselectable="true"
        class="min-w-full h-auto table-auto w-full"
        id="react-aria-:r0:"
        role="grid"
        tabindex="0"
      >
        <thead
          class="[&>tr]:first:rounded-lg"
          role="rowgroup"
        >
          <tr
            class="group outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 cursor-default"
            role="row"
          >
            <th
              class="group px-3 h-10 align-middle bg-default-100 whitespace-nowrap text-foreground-500 text-tiny font-semibold first:rounded-s-lg last:rounded-e-lg data-[sortable=true]:cursor-pointer data-[hover=true]:text-foreground-400 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 text-start"
              data-key="row-header-column-941aw7axa37"
              id="react-aria-:r0:-row-header-column-941aw7axa37"
              role="columnheader"
              tabindex="-1"
            >
              <label
                aria-label="Select All"
                class="group relative max-w-fit inline-flex items-center justify-start cursor-pointer tap-highlight-transparent p-2 -m-2 select-none opacity-disabled pointer-events-none"
                data-disabled="true"
              >
                <span
                  style="border: 0px; clip-path: inset(50%); height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
                >
                  <input
                    aria-label="Select All"
                    aria-labelledby=":r2:"
                    disabled=""
                    type="checkbox"
                    value=""
                  />
                </span>
                <span
                  aria-hidden="true"
                  class="relative inline-flex items-center justify-center flex-shrink-0 overflow-hidden before:content-[''] before:absolute before:inset-0 before:border-solid before:border-2 before:box-border before:border-default after:content-[''] after:absolute after:inset-0 after:scale-50 after:opacity-0 after:origin-center group-data-[selected=true]:after:scale-100 group-data-[selected=true]:after:opacity-100 group-data-[hover=true]:before:bg-default-100 outline-none group-data-[focus-visible=true]:z-10 group-data-[focus-visible=true]:ring-2 group-data-[focus-visible=true]:ring-focus group-data-[focus-visible=true]:ring-offset-2 group-data-[focus-visible=true]:ring-offset-background after:bg-default after:text-default-foreground text-default-foreground w-5 h-5 mr-2 rtl:ml-2 rtl:mr-[unset] rounded-[calc(theme(borderRadius.medium)*0.6)] before:rounded-[calc(theme(borderRadius.medium)*0.6)] after:rounded-[calc(theme(borderRadius.medium)*0.6)] before:transition-colors group-data-[pressed=true]:scale-95 transition-transform after:transition-transform-opacity after:!ease-linear after:!duration-200 motion-reduce:transition-none"
                >
                  <svg
                    aria-hidden="true"
                    class="z-10 opacity-0 group-data-[selected=true]:opacity-100 w-4 h-3 transition-opacity motion-reduce:transition-none"
                    role="presentation"
                    viewBox="0 0 17 18"
                  >
                    <polyline
                      fill="none"
                      points="1 9 7 14 15 4"
                      stroke="currentColor"
                      stroke-dasharray="22"
                      stroke-dashoffset="66"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-width="2"
                    />
                  </svg>
                </span>
              </label>
            </th>
            <th
              class="group px-3 h-10 align-middle bg-default-100 whitespace-nowrap text-foreground-500 text-tiny font-semibold first:rounded-s-lg last:rounded-e-lg data-[sortable=true]:cursor-pointer data-[hover=true]:text-foreground-400 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 text-start w-20"
              data-key="no"
              id="react-aria-:r0:-no"
              role="columnheader"
              tabindex="-1"
            >
              No
            </th>
            <th
              class="group px-3 h-10 align-middle bg-default-100 whitespace-nowrap text-foreground-500 text-tiny font-semibold first:rounded-s-lg last:rounded-e-lg data-[sortable=true]:cursor-pointer data-[hover=true]:text-foreground-400 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 text-start w-16"
              data-key="img"
              id="react-aria-:r0:-img"
              role="columnheader"
              tabindex="-1"
            />
            <th
              class="group px-3 h-10 align-middle bg-default-100 whitespace-nowrap text-foreground-500 text-tiny font-semibold first:rounded-s-lg last:rounded-e-lg data-[sortable=true]:cursor-pointer data-[hover=true]:text-foreground-400 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 text-start"
              data-key="name"
              id="react-aria-:r0:-name"
              role="columnheader"
              tabindex="-1"
            >
              Name
            </th>
            <th
              class="group px-3 h-10 align-middle bg-default-100 whitespace-nowrap text-foreground-500 text-tiny font-semibold first:rounded-s-lg last:rounded-e-lg data-[sortable=true]:cursor-pointer data-[hover=true]:text-foreground-400 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 text-start"
              data-key="status"
              id="react-aria-:r0:-status"
              role="columnheader"
              tabindex="-1"
            >
              Status
            </th>
            <th
              class="group px-3 h-10 align-middle bg-default-100 whitespace-nowrap text-foreground-500 text-tiny font-semibold first:rounded-s-lg last:rounded-e-lg data-[sortable=true]:cursor-pointer data-[hover=true]:text-foreground-400 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 text-start"
              data-key="views"
              id="react-aria-:r0:-views"
              role="columnheader"
              tabindex="-1"
            >
              Views
            </th>
            <th
              class="group px-3 h-10 align-middle bg-default-100 whitespace-nowrap text-foreground-500 text-tiny font-semibold first:rounded-s-lg last:rounded-e-lg data-[sortable=true]:cursor-pointer data-[hover=true]:text-foreground-400 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 text-start"
              data-key="sales"
              id="react-aria-:r0:-sales"
              role="columnheader"
              tabindex="-1"
            >
              Sales
            </th>
            <th
              class="group px-3 h-10 align-middle bg-default-100 whitespace-nowrap text-foreground-500 text-tiny font-semibold first:rounded-s-lg last:rounded-e-lg data-[sortable=true]:cursor-pointer data-[hover=true]:text-foreground-400 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 text-start"
              data-key="conversion"
              id="react-aria-:r0:-conversion"
              role="columnheader"
              tabindex="-1"
            >
              Conversion
            </th>
            <th
              class="group px-3 h-10 align-middle bg-default-100 whitespace-nowrap text-foreground-500 text-tiny font-semibold first:rounded-s-lg last:rounded-e-lg data-[sortable=true]:cursor-pointer data-[hover=true]:text-foreground-400 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 text-start"
              data-key="total"
              id="react-aria-:r0:-total"
              role="columnheader"
              tabindex="-1"
            >
              Total
            </th>
            <th
              class="group px-3 h-10 align-middle bg-default-100 whitespace-nowrap text-foreground-500 text-tiny font-semibold first:rounded-s-lg last:rounded-e-lg data-[sortable=true]:cursor-pointer data-[hover=true]:text-foreground-400 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 text-start"
              data-key="details"
              id="react-aria-:r0:-details"
              role="columnheader"
              tabindex="-1"
            >
              Details
            </th>
          </tr>
          <tr
            aria-hidden="true"
            class="w-px h-px block"
            style="margin-left: 0.25rem; margin-top: 0.25rem;"
            tabindex="-1"
          />
        </thead>
        <tbody
          data-empty="true"
          role="rowgroup"
        >
          <tr
            role="row"
          >
            <td
              class="text-foreground-400 align-middle text-center h-40"
              colspan="10"
              role="gridcell"
            >
              No data found.
            </td>
          </tr>
        </tbody>
      </table>
      <div
        class="flex justify-between items-center mt-4"
      >
        <button
          class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap subpixel-antialiased overflow-hidden tap-highlight-transparent data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 min-w-20 h-10 [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none data-[hover=true]:opacity-hover text-black bg-white dark:border-gray-700 dark:bg-dark-900 dark:text-white rounded-md p-3 gap-4 font-semibold text-base border-1 border-translucent-100"
          type="button"
        >
          <svg
            fill="none"
            height="14px"
            viewBox="0 0 20 15"
            width="19px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              class="fill-dark-800 dark:fill-white"
              clip-rule="evenodd"
              d="M6.92481 13.8671L0.724792 7.68506C0.428955 7.38574 0.428955 6.9043 0.724792 6.60498L6.92481 0.424072C7.11755 0.230713 7.39886 0.154785 7.66278 0.224976C7.92667 0.295166 8.13309 0.500732 8.20428 0.764404C8.27545 1.02808 8.20056 1.30969 8.00781 1.50317L3.1178 6.3811H18.6248C18.8991 6.37891 19.1536 6.52405 19.2914 6.76135C19.4292 6.99854 19.4292 7.2915 19.2914 7.52869C19.1536 7.76599 18.8991 7.91113 18.6248 7.90906H3.11682L8.00684 12.7871C8.30267 13.0864 8.30267 13.5679 8.00684 13.8672C7.7077 14.1663 7.2229 14.1661 6.92383 13.8672L6.92481 13.8671Z"
              fill-rule="evenodd"
            />
          </svg>
          Previous
        </button>
        <span
          class="text-grayBlue-400 font-semibold"
        >
          Page 
          10
           of 
          1
        </span>
        <button
          class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap subpixel-antialiased overflow-hidden tap-highlight-transparent data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 min-w-20 h-10 [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none data-[hover=true]:opacity-hover text-black bg-white dark:border-gray-700 dark:bg-dark-900 dark:text-white rounded-md p-3 gap-4 font-semibold text-base border-1 border-translucent-100"
          type="button"
        >
          Next
          <svg
            fill="none"
            height="14px"
            viewBox="0 0 20 15"
            width="19px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              class="fill-dark-800 dark:fill-white"
              clip-rule="evenodd"
              d="M11.9924 13.8666C11.6967 13.5674 11.6967 13.0859 11.9924 12.7866L16.8824 7.90857H1.37549C1.10107 7.91064 0.84668 7.7655 0.708862 7.5282C0.571045 7.29102 0.571045 6.99805 0.708862 6.76086C0.84668 6.52356 1.10107 6.37842 1.37537 6.38049L16.8824 6.38061L11.9924 1.50256C11.6946 1.20386 11.6953 0.720093 11.9941 0.422241C12.2928 0.124268 12.7766 0.125 13.0745 0.423706L19.2744 6.60461C19.5708 6.90356 19.5708 7.3855 19.2744 7.68457L13.0745 13.8666C12.931 14.01 12.7363 14.0907 12.5334 14.0907C12.3306 14.0907 12.1359 14.01 11.9924 13.8666Z"
              fill-rule="evenodd"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
